Samples/NetworkMediaStreamer/CameraRTPStreamerApp/CameraRTPStreamerApp.cpp

```diff
@@ -0,0 +1,430 @@
+
+            auto serverCerts = getServerCertificate();
+            check_hresult(CreateRTSPServer(streamers.as<ABI::RTSPSuffixSinkMap>().get(), SecureServerPort, true, spBasicAuthProvider.get(), serverCerts.data(), serverCerts.size(), serverHandleSecure.put()));
+        }
+        catch (hresult_error const& ex)
+        {
+            std::wcout << L"\nError starting secure server: " << std::hex << ex.code() << std::dec << L":" << ex.message().c_str();
+            serverHandleSecure = nullptr;
+        }
+#endif
+        auto loggerdelegate = winrt::LogHandler([&fileLogger](auto er, auto msg)
+            {
+                fileLogger << msg.c_str() << "\n";
+                if (er)
+                {
+                    fileLogger << L" ErrCode:" << std::hex << er << "\n" << std::dec;
+                    std::wcout << msg.c_str();
+                    std::wcout << L" ErrCode:" << std::hex << er << "\n" << std::dec;
+                }
+                fileLogger.flush();
+
+            });
+        for (int i = 0; i < (int)LoggerType::LOGGER_MAX; i++)
+        {
+            EventRegistrationToken t1, t2;
+            winrt::check_hresult(serverHandle->AddLogHandler((LoggerType)i, loggerdelegate.as<ABI::LogHandler>().get(), &t1));
+            serverHandleSecure ? winrt::check_hresult(serverHandleSecure->AddLogHandler((LoggerType)i, loggerdelegate.as<ABI::LogHandler>().get(), &t2)) : void(0);
+        }
+        winrt::check_hresult(serverHandle->StartServer());
+        serverHandleSecure ? winrt::check_hresult(serverHandleSecure->StartServer()) : void(0);
+#ifdef USE_FR
+        auto fsources = mc.FrameSources();
+        MediaFrameSource selectedFs(nullptr);
+        std::vector<hstring> preferredsubTypes = { L"NV12", L"YUY2", L"IYUV", L"ARGB32" };
+        for (auto&& fs : fsources)
+        {
+            if (fs.Value().Controller().VideoDeviceController().Id() == s.VideoDeviceId())
+            {
+                selectedFs = fs.Value();
+            }
+        }
+        auto fr = mc.CreateFrameReaderAsync(selectedFs, selectedFs.CurrentFormat().Subtype(), sz).get();
+        fr.AcquisitionMode(MediaFrameReaderAcquisitionMode::Realtime);
+        slim_mutex m;
+        std::vector<com_ptr<IMFSinkWriter>> spSinkWriters;
+        auto strmIter = streamers.First();
+        for (uint32_t i = 0; i < streamers.Size(); i++)
+        {
+            winrt::com_ptr<IMFSinkWriter> spSW;
+            spSW.attach(InitSinkWriter(strmIter.Current().Value().as<IMFMediaSink>().get(), selectedFs.CurrentFormat()));
+            spSinkWriters.push_back(spSW);
+            strmIter.MoveNext();
+        }
+        fr.FrameArrived([&](MediaFrameReader mfr, MediaFrameArrivedEventArgs args)
+            {
+                slim_lock_guard g(m);
+                auto mf = mfr.TryAcquireLatestFrame();
+                if (!mf) return;
+                auto vmf = mf.VideoMediaFrame();
+                if (!vmf) return;

```

Comment:
you can get IMFSample directly from framereference 
auto spGetService= mf.as<IMFGetservice>();
wil::com_ptr<IMFSample> spSample;
spGetService->GetService(MF_WRAPPED_SAMPLE_SERVICE, IID_PPV_ARGS(&spSample));

Reply:
i did not know that. I shall try replacing with this logic.

---
In reply to: [493964653](https://github.com/microsoft/Windows-Camera/pull/13#discussion_r493964653) [](ancestors = 493964653)